storage_proxy: cas: reject for tablets-enabled tables

bhalevy opened the pull request:
Currently, LWT is not supported with tablets.
In particular the interaction between paxos and tablet migration is not handled yet.

Therefore, it is better to outright reject LWT queries for tablets-enabled tables rather than support them in a flaky way.

Fixes scylladb/scylladb#18066

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7798/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7798/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8795/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/768/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7798/)
#### Failed Tests (4653/30230):
- [alternator.run.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7798/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.debug.1)
- [alternator.run.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7798/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.debug.2)
- [alternator.run.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7798/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.debug.3)
- [cql-pytest.cassandra_tests/validation/entities/secondary_index_test.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7798/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_cassandra_tests_validation_entities_secondary_index_test_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.cassandra_tests/validation/entities/secondary_index_test.debug.1)
- [cql-pytest.cassandra_tests/validation/entities/secondary_index_test.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7798/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_cassandra_tests_validation_entities_secondary_index_test_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.cassandra_tests/validation/entities/secondary_index_test.debug.2)
- [cql-pytest.cassandra_tests/validation/entities/secondary_index_test.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7798/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_cassandra_tests_validation_entities_secondary_index_test_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.cassandra_tests/validation/entities/secondary_index_test.debug.3)
- [cql-pytest.cassandra_tests/validation/entities/timestamp_test.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7798/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_cassandra_tests_validation_entities_timestamp_test_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.cassandra_tests/validation/entities/timestamp_test.debug.1)
- [cql-pytest.cassandra_tests/validation/entities/timestamp_test.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7798/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_cassandra_tests_validation_entities_timestamp_test_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.cassandra_tests/validation/entities/timestamp_test.debug.2)
- [cql-pytest.cassandra_tests/validation/entities/timestamp_test.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7798/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_cassandra_tests_validation_entities_timestamp_test_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.cassandra_tests/validation/entities/timestamp_test.debug.3)
- [cql-pytest.cassandra_tests/validation/operations/compact_storage_test.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7798/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_cassandra_tests_validation_operations_compact_storage_test_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.cassandra_tests/validation/operations/compact_storage_test.debug.1)
#### Build Details:
- Duration: 2 hr 38 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
marked unit tests that rely on LWT (indirectly most if not all alternator tests) with xfail_tablets, and made the fixture actually work by yielding a value (it's ignored otherwise, not sure why).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7806/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7806/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8804/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/777/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7806/)
#### Failed Tests (36/30198):
- [alternator.run.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7806/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.debug.1)
- [alternator.run.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7806/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.debug.2)
- [alternator.run.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7806/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.debug.3)
- [alternator.run.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7806/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.dev.1)
- [alternator.run.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7806/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.dev.2)
- [alternator.run.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7806/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.dev.3)
- [alternator.run.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7806/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.release.1)
- [alternator.run.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7806/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.release.2)
- [alternator.run.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7806/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_run_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.run.release.3)
- [test_authorization_no_whitespace](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7806/testReport/junit/%28root%29/test_authorization/Tests___Unit_Tests___test_authorization_no_whitespace) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_authorization_no_whitespace)
#### Build Details:
- Duration: 2 hr 32 min
- Builder: spider5.cloudius-systems.com

bhalevy commented:
In v3 ([cbd13e3](https://github.com/scylladb/scylladb/commit/cbd13e3f54873c618f7cd8c2061edeee2f9da2f9)):
- rebased
- added xfail_tablets also to `test_authorization_no_whitespace`

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7810/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7810/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8807/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/780/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7810/)
#### Build Details:
- Duration: 2 hr 36 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
@nyh please review

nyh reviewed (changes requested):
I think this patch does the wrong thing.

First, for CQL, if I understand correctly it returns the wrong exception when "IF EXISTS" et al. is used - it returns an "internal server error" but should return an orderly Invalid Request error.

More importantly (for me...) is does the wrong thing for Alternator. After https://github.com/scylladb/scylladb/pull/16900, Alternator uses tablets by default on CreateTable (when tablets are available). The correct fix is to change this default - Alternator's default should be NOT to use tablets on a new table. If I understand correctly, your patch leaves the default to use tablets and then each and every write will fail. This doesn't make sense - nobody will be able to use Alternator (with LWT enabled) at all in 6.0 this way...

The fact you needed to hack all the fixtures in the Alternator tests to not run on tablets (and basically, EVERYTHING will fail with tablets enabled!) is a clear indication that this is not the right way :-(

nyh commented:
> > I think this patch does the wrong thing.
> > First, for CQL, if I understand correctly it returns the wrong exception when "IF EXISTS" et al. is used - it returns an "internal server error" but should return an orderly Invalid Request error.
> 
> But the request is valid at the CQL level. It's just not currently supported on the server side.

In my opinion, a user trying this query should get an un-retryable query error, like "INVALID" (I vaguely remember we also had some "unsupported" error?), not a retryable internal-server-error.

> > BTW, I think that the following change is still needed, regardless, otherwise I couldn't use xfail_tablets (I don't think it's currently used anywhere)
> 
> ```diff
> -@pytest.fixture(scope="function")
> +@pytest.fixture(scope="session")
>  def xfail_tablets(request, has_tablets):
>      if has_tablets:
>          request.node.add_marker(pytest.mark.xfail(reason='Test expected to fail when Alternator tables use tablets'))
> +        yield True
> +    else:
> +        yield False
> ```

Yes, I didn't use xfail_tablets anywhere, so it's not surprising it was buggy.
But, I was pretty sure I did test it when writing it... Are you sure yielding a value is necessary in a fixture? Can't it return nothing, like any function?

bhalevy commented:
> Yes, I didn't use xfail_tablets anywhere, so it's not surprising it was buggy.
> But, I was pretty sure I did test it when writing it... Are you sure yielding a value is necessary in a fixture? Can't it return nothing, like any function?

I'm not sure why it was needed for the alternator tests.
FWIW, yielding a value is not needed in cql-pytest.
Anyhow, next version doesn't need this since alternator tests pass when LWT is rejected, with c24bc3b57a0ffc4badebdae4897fdddf507a4518

bhalevy commented:
In v4 ([d058ce6](https://github.com/scylladb/scylladb/commit/d058ce6e19bb840acba52f5ab6f463e81b1f1ca4)):
- throw `invalid_request_exception`
  - unit tests adjusted to expect it
- no need to adapt alternator tests, due to c24bc3b57a0ffc4badebdae4897fdddf507a4518

nyh commented on test/cql-pytest/conftest.py:
```diff
@@ -262,3 +262,8 @@ def has_tablets(cql):
         pytest.skip("Test needs tablets experimental feature on")
+
+@pytest.fixture(scope="function")
+def xfail_tablets(request, has_tablets):
```
The xfail_tablets fixture which you're re-adding here previously existed, but was **deliberately** deleted in commit 7888b23e9e71b42adba1da02b3be795ea73bf834 by @denesb. 

The problem with xfail_tablets was that it loses test converage: This test is run only with tablets (the default), and then skipped - so we lose the ability to catch regression. @avikivity didn't like this. Botond's alternative solution was (e.g., see 3527d0aaedc8bc67b195e230741de6167942f981) to use "parameterize" to run this test twice with and without tablets, and mark the tablets run as xfail and the vnode run without xfail.

nyh commented on test/cql-pytest/test_lwt.py:
```diff
@@ -27,7 +27,7 @@ def table1(cql, test_keyspace):
 # exist, Scylla still needs to read the static column, but forgets to do so.
 # this test reproduces this issue.
-def test_lwt_missing_row_with_static(cql, table1):
+def test_lwt_missing_row_with_static(cql, table1, xfail_tablets): # LWT is not supported with tablets yet. See #18066
```
Unless I'm missing something, all tests in test_lwt.py require vnodes. The simplest solution would be to just change the definition of table1 to create a table explicitly with vnodes - and all the tests will use it. This hack should be temporary (and refer to #18066) but it can solve you the need to do a dozen changes in this file, and also avoid the lowering test coverage (basically, your patch removes all the LWT tests in the cql-pytest framework).

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -201,3 +201,16 @@ def test_tablets_are_dropped_when_dropping_index(cql, test_keyspace, drop_index)
+
+# Test that a LWT query fails with a tablets-enabled table.
+# See issue #18068
+def test_lwt_support_with_tablets(cql, test_keyspace, skip_without_tablets):
```
Good to test this, but please state more clearly that this is a temporary situation that LWT fails with tablets, and when 18068 is solved, we need to delete this test.

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -201,3 +201,16 @@ def test_tablets_are_dropped_when_dropping_index(cql, test_keyspace, drop_index)
+    with new_test_table(cql, test_keyspace, "key int PRIMARY KEY, val int") as table:
+        cql.execute(f"INSERT INTO {table} (key, val) VALUES(1, 0)")
+        with pytest.raises(InvalidRequest, match=f"{table}.*LWT is not yet supported with tablets"):
+            cql.execute(f"INSERT INTO {table} (key, val) VALUES(1, 1) IF NOT EXISTS")
```
Please also check the prepared statement case. If I understand correctly, the prepare() will (unfortunately) work - but then the execute() will fail with InvalidRequest (which is very strange, so let's verify that it actually works).

bhalevy replied:
apologies, I missed it

bhalevy replied:
added in v6

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7981/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7981/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8974/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/946/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7981/)
#### Build Details:
- Duration: 2 hr 7 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
In v5 ([328689a](https://github.com/scylladb/scylladb/commit/328689a0e97d8d7ef2ec1f70b92d6bb2097c98dc)):
- dropped xfail_tablets
- applied @nyh's review comments:
  - use `@pytest.mark.parametrize` to clearly xfail some test with tablets and test them also with vnodes
  - convert tablet fixtures for lwt tests to use test_keyspace_vnodes

bhalevy commented:
@nyh please re-review

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7985/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7985/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/953/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8981/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7985/)
#### Build Details:
- Duration: 2 hr 17 min
- Builder: spider5.cloudius-systems.com

mykaul commented:
@nyh - can you re-review?

nyh reviewed (approved):
This PR doesn't make me happy (see specific comments) but maybe if we want to rush forward with the tablets-that-don't-support-LWT, we have no choice, and this PR is the least-worst way we can do it, so I approve....

bhalevy commented on test/cql-pytest/test_unset.py:
```diff
@@ -201,7 +201,11 @@ def test_unset_insert_where(cql, table2):
+                         [pytest.param("tablets", marks=[pytest.mark.xfail(reason="issue #18066")]), "vnodes"],
+                         indirect=True)
+def test_unset_insert_where_lwt(cql, test_keyspace):
+  with new_test_table(cql, test_keyspace, "p int, c int, PRIMARY KEY (p, c)") as table2:
```
makes sense.  I'll a respective comment.

bhalevy replied:
added comments in v6

bhalevy commented:
In v6 ([f958db6](https://github.com/scylladb/scylladb/commit/f958db6c64ff18acb736d2e2fe88f0a95f614aab)):
- rebased
- added comments about tablets-specific changes to test_unset
- added a test for prepared lw statement to test_lwt_support_with_tablets

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8501/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8501/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_lwt
:small_blue_diamond: cql-pytest/test_non_deterministic_functions
:small_blue_diamond: cql-pytest/test_tablets
:small_blue_diamond: cql-pytest/test_unset
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8501/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1354/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9377/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8501/)
#### Build Details:
- Duration: 3 hr 3 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8501/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8501/) **The following new/updated tests ran 100 times for each mode:** 🔹 cql-pytest/test_lwt 🔹 cql-pytest/test_non_deterministic_functions 🔹 cql-pytest/test_tablets 🔹 cql-pytest/test_unset ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8501/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1354/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9377/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8501/)
> 
> #### Build Details:
> * Duration: 3 hr 3 min
> * Builder: spider2.cloudius-systems.com

https://github.com/scylladb/scylla-dtest/issues/3686

bhalevy commented:
@nyh please re-review.

nyh commented:
> @nyh please re-review.

Even though you "excused" a dtest failure, we still need to run the CI again to see it pass at least once. I'll do it now.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8577/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8577/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_lwt
:small_blue_diamond: cql-pytest/test_non_deterministic_functions
:small_blue_diamond: cql-pytest/test_tablets
:small_blue_diamond: cql-pytest/test_unset
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8577/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9435/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1412/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8577/)
#### Build Details:
- Duration: 5 hr 30 min
- Builder: i-083ec53c58dedf05a (m5d.8xlarge)

nyh commented:
@bhalevy @denesb please note that after this PR, the warning message on CREATE KEYSPACE (added in https://github.com/scylladb/scylladb/pull/17318) is no longer accurate. The warning message says (cql3/statements/create_keyspace_statement.cc):
```c++
            warnings.push_back(
                "Tables in this keyspace will be replicated using tablets, "
                "and will not support the CDC feature (issue #16317) and LWT "
                "may suffer from issue #5251 more often. If you want to use "
                "CDC or LWT, please drop this keyspace and re-create it "
                "without tablets, by adding AND TABLETS = {'enabled': false} "
                "to the CREATE KEYSPACE statement.");
```

The situation today isn't that "LWT may suffer" with tablets - but that it is outright not supported, just like CDC, and this should be reflected by the warning message more accurately.

The fact that LWT isn't supported surprised a user on stackoverflow: https://stackoverflow.com/questions/78623631/scylladb-error-lwt-is-not-yet-supported-with-tablets/78628609#78628609

Also, we **should have** printed either a relevant issue number or the solution or the CREATE KEYSPACE trick, to the error message printed in this PR :-( The stackoverflow user got the error message " LWT is not yet supported with tablets" and had no idea what this means - he didn't know what tablets were, or why he was using them or if there's a way not to use them.

nyh commented:
@bhalevy sorry, false alarm, it appears that this file wasn't actually in the repository, it was just in my working directory ;-) Sorry about the spam.
